src/Handler/UidHandler.php

```diff
@@ -0,0 +1,179 @@
+<?php
+
+declare(strict_types=1);
+
+namespace JMS\Serializer\Handler;
+
+use JMS\Serializer\DeserializationContext;
+use JMS\Serializer\Exception\InvalidArgumentException;
+use JMS\Serializer\GraphNavigatorInterface;
+use JMS\Serializer\SerializationContext;
+use JMS\Serializer\Visitor\DeserializationVisitorInterface;
+use JMS\Serializer\Visitor\SerializationVisitorInterface;
+use JMS\Serializer\XmlSerializationVisitor;
+use Symfony\Component\Uid\AbstractUid;
+use Symfony\Component\Uid\Ulid;
+use Symfony\Component\Uid\Uuid;
+use Symfony\Component\Uid\UuidV1;
+use Symfony\Component\Uid\UuidV3;
+use Symfony\Component\Uid\UuidV4;
+use Symfony\Component\Uid\UuidV5;
+use Symfony\Component\Uid\UuidV6;
+
+final class UidHandler implements SubscribingHandlerInterface
+{
+    public const FORMAT_BASE32    = 'base32';
+    public const FORMAT_BASE58    = 'base58';
+    public const FORMAT_CANONICAL = 'canonical';
+    public const FORMAT_RFC4122   = 'rfc4122';
+
+    private const UID_CLASSES = [
+        Ulid::class,
+        UuidV1::class,
+        UuidV3::class,
+        UuidV4::class,
+        UuidV5::class,
+        UuidV6::class,
+    ];
+
+    /**
+     * @var string
+     * @phpstan-var self::FORMAT_*
+     */
+    private $defaultFormat;
+
+    /**
+     * @var bool
+     */
+    private $xmlCData;
+
+    public function __construct(string $defaultFormat = self::FORMAT_CANONICAL, bool $xmlCData = true)
+    {
+        $this->defaultFormat = $defaultFormat;
+        $this->xmlCData = $xmlCData;
+    }
+
+    public static function getSubscribingMethods(): array
+    {
+        $methods = [];
+        $formats = ['json', 'xml'];
+
+        foreach ($formats as $format) {
+            $methods[] = [
+                'direction' => GraphNavigatorInterface::DIRECTION_SERIALIZATION,
+                'format'    => $format,
+                'type'      => AbstractUid::class,
```

Comment:
could you please add a "integration" test in `tests/Serializer/BaseSerializationTest.php` to makes sure that https://github.com/schmittjoh/serializer/blob/e5d8e8ebccd1f00d04d350dfb30c88b066390cb3/src/GraphNavigator/SerializationGraphNavigator.php#L190 does not interfere with it

Reply:
I'm trying the patch below.  The serialization seems to work right, but it's failing hard on deserializing.  Any ideas?

```sh
TypeError : JMS\Serializer\Handler\SymfonyUidHandler::deserializeUidFromJson(): Argument #2 ($data) must be of type ?string, array given, called in /Users/mbabker/projects/serializer/src/GraphNavigator/DeserializationGraphNavigator.php on line 161
 /Users/mbabker/projects/serializer/src/Handler/SymfonyUidHandler.php:84
 /Users/mbabker/projects/serializer/src/GraphNavigator/DeserializationGraphNavigator.php:161
 /Users/mbabker/projects/serializer/src/Serializer.php:252
 /Users/mbabker/projects/serializer/src/Serializer.php:180
 /Users/mbabker/projects/serializer/tests/Serializer/BaseSerializationTest.php:2005
 /Users/mbabker/projects/serializer/tests/Serializer/BaseSerializationTest.php:936
```

```diff
diff --git a/tests/Serializer/BaseSerializationTest.php b/tests/Serializer/BaseSerializationTest.php
index eb6a0720..16120beb 100644
--- a/tests/Serializer/BaseSerializationTest.php
+++ b/tests/Serializer/BaseSerializationTest.php
@@ -133,6 +133,8 @@ use Symfony\Component\Form\FormConfigBuilder;
 use Symfony\Component\Form\FormError;
 use Symfony\Component\Form\FormFactoryBuilder;
 use Symfony\Component\Translation\IdentityTranslator;
+use Symfony\Component\Uid\Uuid;
+use Symfony\Component\Uid\UuidV4;
 use Symfony\Component\Validator\ConstraintViolation;
 use Symfony\Component\Validator\ConstraintViolationList;
 
@@ -924,6 +926,20 @@ abstract class BaseSerializationTest extends TestCase
         }
     }
 
+    public function testSymfonyUid()
+    {
+        $uid = Uuid::fromString('66b3177c-e03b-4a22-9dee-ddd7d37a04d5');
+
+        self::assertEquals($this->getContent('uid'), $this->serialize($uid));
+
+        if ($this->hasDeserializer()) {
+            $deserialized = $this->deserialize($this->getContent('uid'), UuidV4::class);
+
+            self::assertInstanceOf(UuidV4::class, $deserialized);
+            self::assertTrue($uid->equals($deserialized));
+        }
+    }
+
     public function testExpressionAuthor()
     {
         $evaluator = new ExpressionEvaluator(new ExpressionLanguage());
diff --git a/tests/Serializer/JsonSerializationTest.php b/tests/Serializer/JsonSerializationTest.php
index 19d70cf9..e8d5c622 100644
--- a/tests/Serializer/JsonSerializationTest.php
+++ b/tests/Serializer/JsonSerializationTest.php
@@ -138,6 +138,7 @@ class JsonSerializationTest extends BaseSerializationTest
             $outputs['typed_props'] = '{"id":1,"role":{"id":5},"vehicle":{"type":"car"},"created":"2010-10-01T00:00:00+00:00","updated":"2011-10-01T00:00:00+00:00","tags":["a","b"]}';
             $outputs['custom_datetimeinterface'] = '{"custom":"2021-09-07"}';
             $outputs['data_integer'] = '{"data":10000}';
+            $outputs['uid'] = '{"uid":"66b3177c-e03b-4a22-9dee-ddd7d37a04d5"}';
         }
 
         if (!isset($outputs[$key])) {
diff --git a/tests/Serializer/xml/uid.xml b/tests/Serializer/xml/uid.xml
new file mode 100644
index 00000000..a892f3a4
--- /dev/null
+++ b/tests/Serializer/xml/uid.xml
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<result>
+  <uid><![CDATA[66b3177c-e03b-4a22-9dee-ddd7d37a04d5]]></uid>
+</result>
```

Reply:
Looking at `'{"uid":"66b3177c-e03b-4a22-9dee-ddd7d37a04d5"}'` - it looks like it represents an object with `uid` property with `Symfony\Component\Uid\Uuid` class - you are trying to deserialize it into `Symfony\Component\Uid\Uuid` object. You might need to add some additional Fixture class to fix it.

Reply:
I got it to work by registering the serialization handler on all of the subclasses instead of only the root class.